Use py3-preferred plistlib API to load service plists.

sheagcraig opened the pull request:
### What does this PR do?
This PR simply creates a conditional for py2 vs. py3 plistlib calls in the mac_utils module. In py3 on macOS, the updated func gets called during a bunch of salt calls; each time the log gets a warning about the deprecation of `plistlib.readPlist`. So, specifically, this PR uses the new API for plistlib (matches the json API) to read plists.

### What issues does this PR fix or reference?
Prevents python 3 from issuing deprecation notices during normal salt execution.

### Previous Behavior
```
# salt-call state.apply
...
[WARNING ] /opt/salt/lib/python3.5/site-packages/salt/utils/mac_utils.py:323: DeprecationWarning: The readPlist function is deprecated, use load() instead
  plist = plistlib.readPlist(true_path)

[WARNING ] /opt/salt/lib/python3.5/site-packages/salt/utils/mac_utils.py:345: DeprecationWarning: Attribute access from plist dicts is deprecated, use d[key] notation instead
  'plist': plist}
...
```

### New Behavior
No warnings issues.

### Tests written?
No

### Commits signed with GPG?
No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

sheagcraig committed:
Use py3-preferred plistlib API to load service plists.

sheagcraig committed:
Use py3-preferred plistlib API to load service plists.

sheagcraig commented:
Also, one thing to note. The `try/except` block this is contained within is there to handle binary plists; if plistlib in python2 can't read a plist file and raises an exception, the code (rightly so) uses the macOS cli to convert the file to XML1 format.

However, in py3, plistlib can now handle binary plists, so this exception code either won't help or (more likely) won't get used. Let me know if you'd like me to also clean up the comments and possibly add some logging to clarify what's going on here.

sheagcraig committed:
Use Salt fopen utility instead of python builtin.

sheagcraig committed:
Use Salt fopen utility instead of python builtin.

sheagcraig commented:
Not sure how to avoid the conditional that codeclimate is complaining about. Please advise!

Ch3LL commented:
@sheagcraig thanks for cleaning these warnings up :) looks like there are some related test failures, if you could please take a look at those:

```
unit.utils.test_mac_utils.MacUtilsTestCase.test_available_services
unit.utils.test_mac_utils.MacUtilsTestCase.test_available_services_broken_symlink
```

sheagcraig commented:
I think I see why this is failing those tests. The tests are patching `plistlib.readPlist` with `mock.patch`. The problem is, I can't just patch `plistlib.load` like other tests in this module are doing, because unlike readPlistFromString/loads, `plistlib.load` expects a file handle, whereas `plistlib.readPlist` expects a string which is a path to a plist file; it then handles opening that file and reading it.

I can see one strategy to avoid this is make a conditional decorator for the test that patches `open` if it's python3, but that seems a little wild to me.

Alternately, I could make reading a plist its own function in the mac_utils module and then update the failing tests to patch it instead of plistlib. This strategy would require writing some tests for the new plist loading/reading function I imagine.

Let me know if you have a preference; and if these are silly, please don't hesitate to suggest something better!

sheagcraig committed:
Fix logic in mac_utils tests mocking.

For testing, `os.walk` is getting patched. However, in py3 at least, the
side_effect is returned as a generator. This causes problems, because
`os.walk` is actually called at least four times during the course of
the `mac_utils._available_services()` call. After the first call the
generator is consumed and then on the subsequent call raises a
`StopIteration` error.

My understanding is that unittest.mock has been updated in python 3.7.1
to automatically reset the side-effect each time the patched object is
called, but since 3.7.1 is not out at the time of this PR, my suspicion
is that it's going to be awhile before it makes its way into Salt.

sheagcraig committed:
Fix additional WS.

rallytime commented:
@sheagcraig can you rebase this against he `HEAD` of develop? There are quite a few commits in here that aren't related to you changes and it would be good to keep this clean. :)

sheagcraig commented:
Haha @rallytime yeah, I really jacked it up by mistake.

I was just trying to figure out how to do that!

rallytime commented:
@sheagcraig Looks good now! 👍 

Thanks!

rallytime commented:
@sheagcraig We need to tweak some of the settings in code climate. I think this is fine as-is. Thanks for checking. :)

rallytime commented:
@Ch3LL and @twangboy Can you take a look at this one?

Ch3LL reviewed (approved):
woohoo! thanks for this

rallytime commented:
Thank you @sheagcraig and welcome to the Salt contributor community!
